### PR TITLE
fix of problems with parameter overrides in model inheritance

### LINKED ISSUE #9108: astropy.modeling returns too many Parameters when defining a new Model as a subclass of another Model
When creating an `astropy.modeling.Model` that is a subclass of another `astropy.modeling.Model` (instead of just being a subclass of a `Fittable1DModel` or `Fittable2DModel`, e.g.) the `Parameters` of both the new Model and the super Model are returned, resulting in `TypeError: evaluate() takes X positional arguments but Y were given`.

MWE to reproduce the error:
```
import astropy
print('astropy version', astropy.__version__)

from astropy.modeling import Fittable2DModel, Parameter


class classmodel(Fittable2DModel):
    f = Parameter(default=1)
    x = Parameter(default=0)
    y = Parameter(default=2)

    def __init__(self, f=f.default, x=x.default, y=y.default):
        print(self.param_names)
        print('class input values', f, x, y)
        super().__init__(f, x, y)

    def evaluate(self):
        pass


class subclassmodel(classmodel):
    f = Parameter(default=3)
    x = Parameter(default=10)
    y = Parameter(default=12)
    h = Parameter(default=5)

    def __init__(self, f=f.default, x=x.default, y=y.default, h=h.default):
        print('subclass input values', f, x, y, h)
        super().__init__(f, x, y)

    def evaluate(self):
        pass


print('class should be: f=1, x=0, y=2')
print('subclass should be: f=3, x=10, y=12, h=5')
a = classmodel()
print('class params + names', a.parameters, a.param_names)
b = subclassmodel()
print('subclass params + names', b.parameters, b.param_names)
```

which gives:

```
astropy version 3.2.1
class should be: f=1, x=0, y=2
subclass should be: f=3, x=10, y=12, h=5
('f', 'x', 'y')
class input values 1 0 2
class params + names [1. 0. 2.] ('f', 'x', 'y')
subclass input values 3 10 12 5
('f', 'x', 'y', 'h')
class input values 3 10 12
subclass params + names [ 3. 10. 12.  5.] ('f', 'x', 'y', 'h')
```
and
```
astropy version 4.0.dev25505
class should be: f=1, x=0, y=2
subclass should be: f=3, x=10, y=12, h=5
('f', 'x', 'y')
class input values 1 0 2
class params + names [1. 0. 2.] ('f', 'x', 'y')
subclass input values 3 10 12 5
('f', 'x', 'y', 'f', 'x', 'y', 'h')
class input values 3 10 12
subclass params + names [ 3. 10. 12.  5.] ('f', 'x', 'y', 'f', 'x', 'y', 'h')
```

I am not sure what changed in the new v4.0 modeling framework updates from last week, but a cursory dig into the code suggests that the extension of `param_names` occurs somewhere in [here](https://github.com/astropy/astropy/blob/984461ff0bf4a000e1c037a26560cdc4689b4854/astropy/modeling/core.py#L114-L120), as `param_names` is length 4 before the loop and length 7 afterwards. My initial feeling is that that's a `CompoundModel` thing which multi-nesting subclasses of `Model`s is tripping up, but I am not an expert on the modeling framework.

See astropy/photutils#921 for a reproduction of a 'proper' failing use case, cc @eteq @larrybradley

### PR DESCRIPTION
**Pull request by perrygreenfield**
Addresses issue raised in #9108.

EDIT: Fix #9108

### CODE REVIEW

**Comment by pllim** [comment]
Thanks, Perry! cc @Onoddil

File: astropy/modeling/tests/test_models.py [hunk 1]
```diff
@@ -734,3 +735,38 @@ def test_tabular1d_inverse():
...
+
+
+class classmodel(FittableModel):
+    f = Parameter(default=1)
+    x = Parameter(default=0)
+    y = Parameter(default=2)
+
+    def __init__(self, f=f.default, x=x.default, y=y.default):
+        print(self.param_names)
+        print('class input values', f, x, y)
```
Comment (nden) [inline astropy/modeling/tests/test_models.py]: Remove `print` statements.

File: astropy/modeling/tests/test_models.py [hunk 2]
```diff
@@ -734,3 +735,38 @@ def test_tabular1d_inverse():
...
+
+
+class subclassmodel(classmodel):
+    f = Parameter(default=3)
+    x = Parameter(default=10)
+    y = Parameter(default=12)
+    h = Parameter(default=5)
+
+    def __init__(self, f=f.default, x=x.default, y=y.default, h=h.default):
+        print('subclass input values', f, x, y, h)
```
Comment (nden) [inline astropy/modeling/tests/test_models.py]: and here

**Comment by nden** [comment]
This appears to solve the immediate  problem. However, if I fix one of the parameters in the subclass, the constraint does not propagate.
```
class subclassmodel(classmodel):
    f = Parameter(default=3, fixed=True)
    x = Parameter(default=10)
    y = Parameter(default=12)
    h = Parameter(default=5)

    def __init__(self, f=f.default, x=x.default, y=y.default, h=h.default):
        super().__init__(f, x, y)

b = subclassmodel()
b.f.fixed                                                                                                   
Out[13]: False
```

I haven't tried other parameter attributes but the tests should probably cover this.

**Comment by perrygreenfield** [comment]
done

**Comment by perrygreenfield** [comment]
@nden regarding parameter attributes, I think I see the problem but I'll need to play with it to confirm. If it is what I think it is, it shouldn't be hard to fix.

**Comment by perrygreenfield** [comment]
@nden Yes, it is what I thought it was (though camouflaged by another mechanism). Should be fixed now (I hope :-)

**Pull request merged**